### PR TITLE
Disable gradient checkpointing during no-grad inference to avoid PyTorch warning

### PR DESCRIPTION
**Pull request by qgallouedec**
This PR adds a context manager to temporarily disable gradient checkpointing when running inference in a `torch.no_grad()` context. This prevents a noisy warning from being raised by PyTorch.

```python
from transformers import AutoModelForCausalLM
import torch

device = "cuda"

model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2-0.5B-Instruct").to(device)
model.train()
model.gradient_checkpointing_enable()
input_ids = torch.tensor([[1, 2, 3]]).to(device)

with torch.no_grad():
    output = model(input_ids=input_ids)
```

When 

1. the model is in train model
2. gradient checkpointing is enabled with `use_reentrant=True` (default)
3. the inference is run within a `torch.no_grad()` context

we get the following warning:

```
/fsx/qgallouedec/miniconda3/envs/trl/lib/python3.12/site-packages/torch/utils/checkpoint.py:85: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn(
```

This is a safe warning, because no gradients are expected to be computed within the `torch.no_grad()` context, but it is noisy. To avoid this warning, we can disable gradient checkpointing before entering the `torch.no_grad()` context, and re-enable it afterwards.

```python
from transformers import AutoModelForCausalLM
import torch
from trl.models.utils import disable_gradient_checkpointing

device = "cuda"

model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2-0.5B-Instruct").to(device)
model.train()
model.gradient_checkpointing_enable()
input_ids = torch.tensor([[1, 2, 3]]).to(device)

with torch.no_grad(), disable_gradient_checkpointing(model):
    output = model(input_ids=input_ids)
```

### CODE REVIEW

File: trl/models/utils.py [hunk 1]
```diff
@@ -522,3 +522,20 @@ def prepare_peft_model(
...
+    Args:
+        model (`PreTrainedModel`):
+            Model for which to temporarily disable gradient checkpointing.
+    """
+    was_enabled = model.is_gradient_checkpointing
+    if was_enabled:
+        model.gradient_checkpointing_disable()
+    yield
+    if was_enabled:
+        model.gradient_checkpointing_enable()
```
Comment (albertvillanova) [inline trl/models/utils.py]: This part of the code will not be called if an exception is raised after the `yield`. I would recommend using a `try...finally` block.

Comment (albertvillanova) [inline trl/models/utils.py]: Just a question: You assume no `gradient_checkpointing_kwargs` was originally passed to enable checkpointing, e.g. `use_reentrant` is `True`. Maybe this assumption is OK with TRL. Just wanted to be sure of this.

File: trl/models/utils.py [hunk 2]
```diff
@@ -522,3 +522,20 @@ def prepare_peft_model(
...
+    Temporarily disable gradient checkpointing, restoring the previous state afterward.
+
+    Args:
+        model (`PreTrainedModel`):
+            Model for which to temporarily disable gradient checkpointing.
+    """
+    was_enabled = model.is_gradient_checkpointing
+    if was_enabled:
+        model.gradient_checkpointing_disable()
+    yield
```
Comment (albertvillanova) [inline trl/models/utils.py]: What about yielding the model? Feel free to ignore this if you think it is OK, but this pattern is widely used 
```python
with func_to_modify_an_instance(instance) as modified_instance:
````

File: trl/models/utils.py (hunk 1 above)
Comment (qgallouedec) [reply trl/models/utils.py]: I yes, I missed this, thanks!

File: trl/models/utils.py (hunk 2 above)
Comment (qgallouedec) [reply trl/models/utils.py]: I think this could be a little misleading, since the operation is done in-place within the context manager. As a reader, I would get the impression that we are getting a new instance for which grad checkpointing is disabled.

File: trl/models/utils.py (hunk 1 above)
Comment (qgallouedec) [reply trl/models/utils.py]: Addressed here https://github.com/huggingface/trl/pull/4636/commits/68bbbc3e7f0025f8dba7f0ce389b49fe34e1cb44

Comment (qgallouedec) [reply trl/models/utils.py]: Addressed here: https://github.com/huggingface/trl/pull/4636/commits/68bbbc3e7f0025f8dba7f0ce389b49fe34e1cb44

**Comment by qgallouedec** [comment]
> Additionally, I have a naive question: what would be the issues if we temporarily set `model.eval()` instead?

`eval` is actually quite different:

| | `model.eval()` | `torch.no_grad()` |
| - | - | - |
| Disables dropout | Yes | No |
| BatchNorm uses running stats | Yes | No |
| Gradient computation | Yes  | No |

(here, we want to disable the gradient computation only.

in short, `eval()` changes layer behavior, `no_grad()` changes autograd behavior.

**Comment by qgallouedec** [comment]
I'll merge this one so that we have it in 0.26
